### PR TITLE
feat(new integration): june cloud mode destination onboarding

### PR DESCRIPTION
**Pull request by Gauravudia**
## Description of the change

> Onboard June Cloud Mode Destination

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/june/data/JuneGroupConfig.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+[
+  {
+    "destKey": "userId",
+    "sourceKeys": "userIdOnly",
+    "sourceFromGenericMap": true,
+    "required": false
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply v0/destinations/june/data/JuneGroupConfig.json:6]: To create a new company userId is required. To update the existing company userId is not required.

File: v0/destinations/june/transform.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+const trackResponseBuilder = (message, destination) => {
+  const { endpoint } = CONFIG_CATEGORIES.TRACK;
+  const { groupId } = message.properties || {};
+  let payload = constructPayload(
+    message,
+    MAPPING_CONFIG[CONFIG_CATEGORIES.TRACK.name]
+  );
+
+  if (groupId) {
+    payload = { ...payload, context: { groupId } };
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply v0/destinations/june/transform.js:48]: if groupId is null/undefined/blank i am not sending the `context` object itself.
context.groupId is optional field.

File: v0/destinations/june/data/JuneIdentifyConfig.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
+[
+  {
+    "destKey": "userId",
+    "sourceKeys": "userIdOnly",
+    "sourceFromGenericMap": true,
+    "required": true
+  },
+  {
+    "destKey": "timestamp",
+    "sourceKeys": "createdAt",
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply v0/destinations/june/data/JuneIdentifyConfig.json]: updated

File: v0/destinations/june/data/JuneTrackConfig.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    "destKey": "properties",
+    "sourceKeys": "properties",
+    "required": false,
+    "metadata": {
+      "excludes": ["groupId"]
+    }
+  },
+  {
+    "destKey": "timestamp",
+    "sourceKeys": "createdAt",
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply v0/destinations/june/data/JuneTrackConfig.json]: updated

File: v0/destinations/june/transform.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gauravudia) [reply v0/destinations/june/transform.js:48]: We can send array/object in traits
Screenshot 2022-09-22 at 11 51 30 AM

File: __tests__/data/june.json [hunk 5]
```diff
@@ -0,0 +1,323 @@
+[
+  {
+    "description": "Identify call without userId",
```
Comment (shrouti1507) [inline __tests__/data/june.json:3]: to remind, we need to add this info in doc. If not mentioned by now

File: __tests__/data/june.json [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+      "message": {
+        "type": "track",
+        "event": "Product Reviewed",
+        "userId": "5136633649",
+        "properties": {
+          "review_id": "12345",
+          "product_id": "123",
+          "rating": 3.0,
+          "review_body": "Average product, expected much more.",
+          "groupId": "91Yb32830"
```
Comment (shrouti1507) [inline __tests__/data/june.json:225]: Need to add a callout in doc, regarding:

- what happens if the group ID is passed through the track call
- If the user wants to pass group ID we are expecting that in properties

File: v0/destinations/june/transform.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  const payload = constructPayload(
+    message,
+    MAPPING_CONFIG[CONFIG_CATEGORIES.IDENTIFY.name]
+  );
+  return responseBuilder(payload, endpoint, destination);
+};
+
+const trackResponseBuilder = (message, destination) => {
+  const { endpoint } = CONFIG_CATEGORIES.TRACK;
+  const { groupId } = message.properties || {};
```
Comment (shrouti1507) [inline v0/destinations/june/transform.js:51]: Lets also add an externalID support. 

groupID = externalId or message.properties.groupId.

File: __tests__/data/june.json (hunk 6 above)
Comment (shrouti1507) [reply __tests__/data/june.json:225]: lets also expect the groupId from externalD as well.
groupId = externalID or properties.groupID

File: __tests__/data/june.json (hunk 5 above)
Comment (Gauravudia) [reply __tests__/data/june.json:3]: userId is required field. Already, mentioned this in research spec.

File: __tests__/data/june.json (hunk 6 above)
Comment (Gauravudia) [reply __tests__/data/june.json:225]: added externalId support

File: v0/destinations/june/transform.js (hunk 7 above)
Comment (Gauravudia) [reply v0/destinations/june/transform.js]: added externalId support

**Review by mihir-4116 — APPROVED** [review]

**Review by shrouti1507 — APPROVED** [review]

**Pull request merged**